eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs

```diff
@@ -276,44 +217,110 @@ impl Eth2NearRelay {
+            } else if mid_block_number > block_number {
+                right_bound = mid_non_empty_slot;
+            } else {
+                left_bound = mid_slot;
+            }
+        }
+
+        return Err(Box::new(SlotByBlockNumberNotFound));
+    }
+
+    fn submit_headers(&mut self) -> bool {
+        info!(target: "relay", "Submit Headers mode");
+        let max_slot = return_val_on_fail!(self.eth_client_contract.get_finalized_beacon_block_slot(),
+            "Fail to get last finalized slot",
+            false
+        );
 
-            if last_eth2_slot_on_near < max_slot_for_submission {
-                info!(target: "relay", "= Creating headers batch =");
+        let min_block_number = return_val_on_fail!(self.eth_client_contract.get_last_block_number(),
+                                               "Fail to get last block number",
+                                               false);
+        let min_slot = return_val_on_fail!(self.get_slot_by_block_number(min_block_number, max_slot, 1),
+            "Fail to get slot by block number",
+            false);
+        let mut current_slot = max_slot;
 
-                let (headers, current_slot) = skip_fail!(
+        info!(target: "relay", "Submit headers from {} to {}", max_slot, min_slot);
+
+        while current_slot > min_slot {
+            info!(target: "relay", "= Creating headers batch =");
+
+            let max_block_number = skip_fail!(self.eth_client_contract.get_unfinalized_tail_block_number(),
+                "Fail to fetch unfinalized tail block number",
+                self.sleep_time_on_sync_secs);
+
+            if let Some(max_block_number) = max_block_number {
+                current_slot = skip_fail!(self.get_slot_by_block_number(max_block_number, current_slot + 1, min_slot), "Fail to get slot by block_number", 0) - 1;
+            }
+
+            let (headers, new_current_slot) = skip_fail!(
                     self.get_execution_blocks_between(
-                        last_eth2_slot_on_near + 1,
-                        max_slot_for_submission,
+                        current_slot,
+                        min_slot,
                     ),
                     "Network problems during fetching execution blocks",
                     self.sleep_time_on_sync_secs
                 );
-                self.submit_execution_blocks(headers, current_slot, &mut last_eth2_slot_on_near);
-                were_submission_on_iter = true;
-            }
-
-            were_submission_on_iter |=
-                self.send_light_client_updates_with_checks(last_eth2_slot_on_near);
 
-            if !were_submission_on_iter {
-                info!(target: "relay", "Sync with ETH network. Sleep {} secs", self.sleep_time_on_sync_secs);
-                thread::sleep(Duration::from_secs(self.sleep_time_on_sync_secs));
+            if !self.submit_execution_blocks(headers, new_current_slot + 1, &mut current_slot) {
```

Comment:
Could we use the execution blocks numbers instead slots hear? 
eg:
```
let mut max_block_number = eth_client_contract.get_unfinalized_tail_block_number();
if max_block_number.is_none() {
    max_block_number = get_block_number_by_slot(eth_client_contract.get_finalized_beacon_block_slot())
}
let min_block_number = eth_client_contract.last_block_number();
```

Reply:
fixed https://github.com/aurora-is-near/rainbow-bridge/pull/883/commits/87ace5cd1c6755116147dc817836ae5d00bf7903